### ISSUE TITLE
so has anyone got this pytorch crf to work yet ?

### ISSUE THREAD

**Issue by yuhujia** [kmkurn/pytorch-crf#50]
Looks like this implementation does not work on real data. If anyone has gotten it to work (aka has good performance and (optional) deployed to production) pls correct me here.

**Comment by kmkurn** [comment]
I have used this myself for production (when I was still working for my former employer). Sorry it doesn't work for you.

**Comment by yuhujia** [comment]
great ! can you post scripts/data/steps to reproduce your performance ?

**Comment by kmkurn** [comment]
Unfortunately no, because it was all proprietary. But it was basically an implementation of [this paper](https://aclweb.org/anthology/C/C16/C16-1030.pdf).

**Comment by yuhujia** [comment]
Steps to reproduce numbers mentioned in the paper would be great.

**Comment by kmkurn** [comment]
I'm not sure what steps you need other than those described in the paper.

**Comment by robinsongh381** [comment]
I have made good products with REAL data :smile:

**Comment by elsheikh21** [comment]
@robinsongh381 @kmkurn @yuhujia 
anyone got it to work or not yet?

**Comment by kmkurn** [comment]
@elsheikh21 Have you tried it yourself? I personally have used this, and it worked.

**Comment by elsheikh21** [comment]
I did and could not use it, may I ask you a question or two?

**Comment by elsheikh21** [comment]
I personally tried to use it, however, I could not. I sent you an email

**Comment by yuye2133** [comment]
There is a problem with the latest version of pytorch(1.4), log-likehihood will overflow.

**Comment by elsheikh21** [comment]
To all future readers, I would like to start by thanking @kmkurn  for helping me out, and here is how to utilize his module, I hope it makes everything clear for you (future reader)

So, how to make this work?
**F**irstly, install the module
`pip install pytorch-crf`
**S**econdly, in your script
`from torchcrf import CRF`

**T**hirdly, add the CRF layer after the model's final layer, 
`self.crf = CRF(num_classes, batch_first=True)` where number of tags/output calsses 

**F**orth, in the training loop

```{python}
for epoch in range(epochs):
   self.model.train()
   for batch_idx, samples in enumerate(train_dataset):
      loss = -self.model.log_probs(inputs, labels)
      sample_loss.backward()
      self.optimizer.step()
   # Compute val_loss
```

But what is model.log_probs(inputs, labels)?
```{python}
    """
     Build a simple BiLSTM 
    """
    def forward(self, x):
        ''' Leave model forward function untouched '''
        embeddings = self.word_embedding(x)  # [Samples_Num, Seq_Len]
        embeddings = self.dropout(embeddings)
        o, _ = self.lstm(embeddings)        # output shape: [Samples_Num, Seq_Len, Tags_Num]
        o = self.dropout(o)
        logits = self.classifier(o)
        return logits

    def log_probs(self, x, tags, mask=None):
        emissions = self(x) # Compute Emission scores
        return self.crf(emissions, tags, mask=mask)
```

**L**astly, how to predict outputs as per model?
`predictions = self.model.predict(inputs)  # returns predicted classes, no need to invoke argmax(..)` 

what is `model.predict(inputs)`?

```{python}
    def predict(self, x):
        emissions = self(x)
        return self.crf.decode(emissions)
```

I was using `Python3.6.4` and `PyTorch 1,4,0`

**Comment by kmkurn** [comment]
The library seems to work just fine, so I'm closing this issue.
